### PR TITLE
Add cbzip reader (and more!)

### PR DESCRIPTION
**Pull request by nickvanw**
* Adds a cbzip reader w/ tests
* Reworks error handling to function properly in the case of errors, as well as `io.EOF` when close and Write are called post-close
* No unique error type

### CODE REVIEW

File: reader.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	r      io.Reader
+	bz     *C.bz_stream
+	in     []byte
+	skipIn bool
+	err    error
+}
+
+// NewBzipWriter returns an io.WriteCloser. Writes to this writer are
+// compressed and sent to the underlying writer.
+// It is the callers responsibility to call Close on the WriteCloser.
```
Comment (mdlayher) [inline reader.go]: Nitpick: "caller's"

File: writer.go [hunk 2]
```diff
@@ -12,57 +12,18 @@ import (
...
+	err error
 }
 
-// NewBzipWriter returns an io.WriteCloser. Writes to this writer are
+// NewWriter returns an io.WriteCloser. Writes to this writer are
 // compressed and sent to the underlying writer.
 // It is the callers responsibility to call Close on the WriteCloser.
 // Writes may not be flushed until Close.
-func NewBzipWriter(w io.Writer) (io.WriteCloser, error) {
+func NewWriter(w io.Writer) (io.WriteCloser, error) {
```
Comment (mdlayher) [inline writer.go]: I have always heard "return a concrete type, but use interfaces for parameters".  If you return an `io.WriteCloser` here, the caller can't use the `Flush()` method without a type-assertion.  In addition, `cgzip` returns a `*Writer` for its `NewWriter()` function as well.

https://github.com/youtube/vitess/blob/master/go/cgzip/writer.go#L83

File: reader.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+type bzipReader struct {
+	r      io.Reader
+	bz     *C.bz_stream
+	in     []byte
+	skipIn bool
+	err    error
+}
+
+// NewBzipWriter returns an io.WriteCloser. Writes to this writer are
```
Comment (mdlayher) [inline reader.go]: This doc comment is for `NewReader()`, not `NewWriter()`.

File: writer.go [hunk 4]
```diff
@@ -12,57 +12,18 @@ import (
...
 	w   io.Writer
 	bz  *C.bz_stream
 	out []byte
+	err error
 }
 
-// NewBzipWriter returns an io.WriteCloser. Writes to this writer are
+// NewWriter returns an io.WriteCloser. Writes to this writer are
 // compressed and sent to the underlying writer.
 // It is the callers responsibility to call Close on the WriteCloser.
```
Comment (mdlayher) [inline writer.go]: Nitpick: "caller's"

File: reader.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+type Reader struct {
+	r      io.Reader
+	bz     *C.bz_stream
+	in     []byte
+	skipIn bool
+	err    error
+}
+
+// NewReader returns an io.Reader. Reads from this are read from the
```
Comment (mdlayher) [inline reader.go]: It's _technically_ an `io.ReadCloser`.  I see you changed the return type to `*Reader` too.  Shouldn't make a big difference either way on this one though.

File: reader.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+// underlying io.Reader and compressed via bzip2 compression
```
Comment (mdlayher) [inline reader.go]: `Reader` decompresses bzip2 compression, right?

File: reader.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	rdr.bz = &C.bz_stream{bzalloc: nil, bzfree: nil, opaque: nil}
+
+	if result := C.BZ2_bzDecompressInit(rdr.bz, verbosity, 0); result != BZ_OK {
+		return nil, ErrInit
+	}
+
+	return rdr, nil
+}
+
+// Read pulls data up from the underlying io.Reader and bzip encodes it
```
Comment (mdlayher) [inline reader.go]: `s/encodes/decodes/`

File: README.md [hunk 8]
```diff
@@ -37,28 +38,65 @@ const (
 copied from bzlib.h
 
 
+## Variables
+``` go
+var (
+    ErrBadParam       = errors.New("bad param (should be impossible)")
+    ErrBadData        = errors.New("integrity problem detected in input data")
+    ErrBadMagic       = errors.New("compressed stream does not being with magic bytes")
+    ErrMem            = errors.New("insufficient memory available ಠ_ಠ")
```
Comment (macb) [inline README.md]: dis guy

**Comment by macb** [comment]
:+1:

**Pull request merged**